### PR TITLE
linux-kernel: Enable BTF

### PR DESCRIPTION
**Pull request by roblabla**
###### Motivation for this change

BTF is a new, lightweight debug information format tailored specifically
for the needs of eBPF, allowing eBPF programs to be portable across
various kernel versions, configurations and distributions. This is used
by bpftrace and lots of new eBPF-based tooling to avoid a dependency
on LLVM on the host.

BTF debug information is enabled on all major distributions: Fedora 31+,
RHEL 8.2+, Ubuntu 20.10, Debian 11 and ArchLinux all have enabled it.

Enabling BTF debug information requires adding two new dependencies to
the kernel build: Python3 and pahole. Those will be used to generate the
BTF debugging information.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by fortuneteller2k** [comment]
@ofborg build linux_xanmod

**Comment by fortuneteller2k** [comment]
@ofborg test latestKernel.login kernel-latest

**Review by moni-dz — APPROVED** [review]

**Comment by Atemu** [comment]
Please target staging, this causes a ton of rebuilds which most users won't directly benefit from any time soon.

**Comment by roblabla** [comment]
How should I do that? Just change the target branch in github's ui?

**Comment by lovesegfault** [comment]
> How should I do that? Just change the target branch in github's ui?

I'm not sure that works. For me I do:
```
git fetch --all
git checkout $myBranch
git rebase -i origin/staging
git push --force
```

You may or may not fiddle with GH's UI as well, not sure.

**Comment by Atemu** [comment]
No, don't do that. That way you'd try to merge staging into master and ping pretty much *all* codeowners.

Just change the base (or target) branch in the GH UI. You only need to rebase if there's a merge conflict between your branch and staging; *after* you changed the target branch that is.

**Comment by roblabla** [comment]
Done, and no conflicts!

**Review by moni-dz — APPROVED** [review]

File: pkgs/os-specific/linux/kernel/manual-config.nix
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/kernel/manual-config.nix:317]: You are sure about .dev in nativeBuildInputs?

File: pkgs/os-specific/linux/kernel/common-config.nix
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/kernel/common-config.nix:39]:
```suggestion
      DEBUG_INFO                = if (features.debug or (whenAtLeast "5.2" yes) or false) then yes else no;
```

File: pkgs/os-specific/linux/kernel/manual-config.nix
Comment (roblabla) [reply pkgs/os-specific/linux/kernel/manual-config.nix:317]: I mean, it works here. I had to specify the .dev package so the headers were available, but maybe it should go in another kind of buildInput? (I'm never 100% clear if packages should go in buildInputs, nativeBuildInputs, or elsewhere tbh).

File: pkgs/os-specific/linux/kernel/common-config.nix
Comment (roblabla) [reply pkgs/os-specific/linux/kernel/common-config.nix]: Did it slightly differently because I couldn't get it to work with `if`s.

File: pkgs/os-specific/linux/kernel/manual-config.nix
Comment (Atemu) [reply pkgs/os-specific/linux/kernel/manual-config.nix:317]: nativeBuildInputs is for dependencies that are supposed to be run during the build; i.e. cmake.

buildInputs is for regular dependencies whose files content is referenced by or during the build; i.e. libraries.

Whether it works or not isn't really the important bit here, it's about declaring things properly so that everything works nicely as it should.  
Cross-compilation is a good example as nativeBuildInputs need to run on the builder which has a different CPU ISA than the target. This is mostly done automatically but the system needs to be told what's what to be able to do such things.

**Comment by markpash** [comment]
Hi, what else needs to be done for this PR to get approvals and get merged? Having BTF info in the kernel is very useful for software packages that make use of eBPF and for eBPF development.

File: pkgs/os-specific/linux/kernel/manual-config.nix
Comment (SuperSandro2000) [reply pkgs/os-specific/linux/kernel/manual-config.nix:317]:
```suggestion
      ++ optionals  (lib.versionAtLeast version "5.2")  [ cpio python3 zlib.dev pahole ]
```

**Comment by lovesegfault** [comment]
If someone resolves the conflict and pings me I will merge this.

**Comment by roblabla** [comment]
Very sorry for the ping, I accidentally rebased on master :sweat_smile: .

@lovesegfault I fixed the merge conflict. I'm still not 100% sure that the zlib.dev dependency is correct though. I'm not familiar enough with the kernel build system or nix to know for sure.

File: pkgs/os-specific/linux/kernel/manual-config.nix
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/kernel/manual-config.nix:319]:
```suggestion
      ++ optionals  (lib.versionAtLeast version "5.2")  [ cpio zlib.dev pahole ]
```

File: pkgs/os-specific/linux/kernel/manual-config.nix
Comment (lovesegfault) [reply pkgs/os-specific/linux/kernel/manual-config.nix]: Yeah, I really don't think `zlib.dev` is the right thing there.

You probably want to add `zlib` to `buildInputs` instead of `zlib.dev` to `nativeBuildInputs`.

**Comment by roblabla** [comment]
Alright, I added zlib in buildInputs, and locally verified that the build works.

**Comment by Mic92** [comment]
Linux 5.10 and linux 5.12 fail to configure with this option.

Please have a look.

```
error: builder for '/nix/store/dn8bf57symyrsna58kvz6nlml76jn3c3-linux-config-5.10.57.drv' failed with exit code 255;                                                                                                
       last 10 log lines:                                                                                                                                                                                           
       > warning: unused option: INET_RAW_DIAG                                                                                                                                                                      
       > warning: unused option: INET_TCP_DIAG                                                                                                                                                                      
       > warning: unused option: INET_UDP_DIAG                                                                                                                                                                      
       > warning: unused option: RCU_PERF_TEST                                                                                                                                                                      
       > warning: unused option: SECURITY_WRITABLE_HOOKS                                                                                                                                                            
       > warning: option not set correctly: STRICT_DEVMEM (wanted 'y', got 'n')
       > warning: unused option: XEN_SCRUB_PAGES
       > warning: unused option: XEN_SELFBALLOONING
       > warning: unused option: XEN_STUB
       > warning: unused option: XEN_TMEM
       For full logs, run 'nix log /nix/store/dn8bf57symyrsna58kvz6nlml76jn3c3-linux-config-5.10.57.drv'.
error: builder for '/nix/store/jgaja973qf0jbs9myjwl29k5r3jh3i2z-linux-config-5.12.19.drv' failed with exit code 255;
       last 10 log lines:
       > warning: unused option: INET_RAW_DIAG
       > warning: unused option: INET_TCP_DIAG
       > warning: unused option: INET_UDP_DIAG
       > warning: unused option: RCU_PERF_TEST
       > warning: unused option: SECURITY_WRITABLE_HOOKS
       > warning: option not set correctly: STRICT_DEVMEM (wanted 'y', got 'n')
       > warning: unused option: XEN_SCRUB_PAGES
       > warning: unused option: XEN_SELFBALLOONING
       > warning: unused option: XEN_STUB
       > warning: unused option: XEN_TMEM
       For full logs, run 'nix log /nix/store/jgaja973qf0jbs9myjwl29k5r3jh3i2z-linux-config-5.12.19.drv'.
```

**Comment by roblabla** [comment]
Well, that shux. It works on my machine, and I tested kernel 5.10 and 5.12. To test, I use a flake.nix system that looks like this:

```
{
    inputs.nixpkgs.url = "github:roblabla/nixpkgs/enable-btf";
    packages.x86_64-linux.kernel = nixpkgs.legacyPackages.x86_64-linux.linuxPackages_5_10.kernel;
    packages.x86_64-linux.configfile = nixpkgs.legacyPackages.x86_64-linux.linuxPackages_5_10.kernel.configfile;
  };
}
```

and then just run `nix build '.#configfile'`. The lockfile pins commit 34150f86c0a6b9d452bab09157896c3b85b103d1 now.

@Mic92 Could you provide the full logs so I can figure out what went wrong?

**Comment by Mic92** [comment]
I think it's only `linuxPackages_5_10_hardened.kernel` now which is sort of broken anyway.

**Comment by lopsided98** [comment]
zlib needs to be in `nativeBuildInputs`, otherwise the kernel fails to cross-compile. zlib appears to only be used for a build system tool, so this makes sense.

**Comment by Izorkin** [comment]
With this PR error build kernel:
```
  LD [M]  drivers/net/wireless/intersil/hostap/hostap_pci.ko
  LD [M]  drivers/net/wireless/intel/iwlwifi/iwlwifi.ko
  LD [M]  drivers/net/wireless/intersil/orinoco/orinoco.ko
/nix/store/fp2h123793bxln4cjxy0mz08sf4nx2m3-binutils-2.35.1/bin/ld: drivers/net/wireless/intel/ipw2x00/ipw2100.ko: final close failed: No space left on device
/nix/store/fp2h123793bxln4cjxy0mz08sf4nx2m3-binutils-2.35.1/bin/ld: drivers/net/wireless/ath/wil6210/wil6210.ko: final close failed: No space left on device
/nix/store/fp2h123793bxln4cjxy0mz08sf4nx2m3-binutils-2.35.1/bin/ld: final link failed: No space left on device
make[3]: *** [../scripts/Makefile.modfinal:78: drivers/net/wireless/intel/ipw2x00/ipw2100.ko] Error 1
make[3]: *** Waiting for unfinished jobs....
make[3]: *** [../scripts/Makefile.modfinal:78: drivers/net/wireless/ath/wil6210/wil6210.ko] Error 1
make[3]: *** [../scripts/Makefile.modfinal:78: drivers/net/wireless/broadcom/b43legacy/b43legacy.ko] Error 1
/nix/store/qfb4j7w2fjjq953nd9xncz5mymj5n0kb-bash-5.1-p8/bin/bash: line 1: printf: write error: No space left on device
make[3]: *** [../scripts/Makefile.modfinal:78: drivers/net/wireless/cisco/airo_cs.ko] Error 1
make[3]: *** Deleting file 'drivers/net/wireless/cisco/airo_cs.ko'
/nix/store/fp2h123793bxln4cjxy0mz08sf4nx2m3-binutils-2.35.1/bin/ld: final link failed: No space left on device
/nix/store/fp2h123793bxln4cjxy0mz08sf4nx2m3-binutils-2.35.1/bin/ld: final link failed: No space left on device
/nix/store/fp2h123793bxln4cjxy0mz08sf4nx2m3-binutils-2.35.1/bin/ld: final link failed: No space left on device
make[3]: *** [../scripts/Makefile.modfinal:78: drivers/net/wireless/intel/ipw2x00/ipw2200.ko] Error 1
make[3]: *** [../scripts/Makefile.modfinal:78: drivers/net/wireless/cisco/airo.ko] Error 1
make[3]: *** [../scripts/Makefile.modfinal:78: drivers/net/wireless/broadcom/brcm80211/brcmsmac/brcmsmac.ko] Error 1
/nix/store/fp2h123793bxln4cjxy0mz08sf4nx2m3-binutils-2.35.1/bin/ld: final link failed: No space left on device
/nix/store/fp2h123793bxln4cjxy0mz08sf4nx2m3-binutils-2.35.1/bin/ld: final link failed: No space left on device
make[3]: *** [../scripts/Makefile.modfinal:78: drivers/net/wireless/intel/iwlegacy/iwl3945.ko] Error 1
/nix/store/fp2h123793bxln4cjxy0mz08sf4nx2m3-binutils-2.35.1/bin/ld: final link failed: No space left on device
make[3]: *** [../scripts/Makefile.modfinal:78: drivers/net/wireless/intel/ipw2x00/libipw.ko] Error 1
/nix/store/fp2h123793bxln4cjxy0mz08sf4nx2m3-binutils-2.35.1/bin/ld: final link failed: No space left on device
make[3]: *** [../scripts/Makefile.modfinal:78: drivers/net/wireless/broadcom/brcm80211/brcmfmac/brcmfmac.ko] Error 1
make[3]: *** [../scripts/Makefile.modfinal:78: drivers/net/wireless/broadcom/b43/b43.ko] Error 1
/nix/store/fp2h123793bxln4cjxy0mz08sf4nx2m3-binutils-2.35.1/bin/ld: drivers/net/wireless/ath/ath9k/ath9k_htc.ko: final close failed: No space left on device
make[3]: *** [../scripts/Makefile.modfinal:78: drivers/net/wireless/ath/ath9k/ath9k_htc.ko] Error 1
/nix/store/fp2h123793bxln4cjxy0mz08sf4nx2m3-binutils-2.35.1/bin/ld: final link failed: No space left on device
make[3]: *** [../scripts/Makefile.modfinal:78: drivers/net/wireless/intersil/hostap/hostap_cs.ko] Error 1
/nix/store/fp2h123793bxln4cjxy0mz08sf4nx2m3-binutils-2.35.1/bin/ld: final link failed: No space left on device
make[3]: *** [../scripts/Makefile.modfinal:78: drivers/net/wireless/ath/ath9k/ath9k_hw.ko] Error 1
/nix/store/fp2h123793bxln4cjxy0mz08sf4nx2m3-binutils-2.35.1/bin/ld/nix/store/fp2h123793bxln4cjxy0mz08sf4nx2m3-binutils-2.35.1/bin/ld: : drivers/net/wireless/intersil/hostap/hostap.kodrivers/net/wireless/intersil/orinoco/orinoco.ko: final close failed: : final close failed: No space left on deviceNo space left on device

make[3]: *** [../scripts/Makefile.modfinal:78: drivers/net/wireless/intersil/hostap/hostap.ko] Error 1
make[3]: *** [../scripts/Makefile.modfinal:78: drivers/net/wireless/intersil/orinoco/orinoco.ko] Error 1
/nix/store/fp2h123793bxln4cjxy0mz08sf4nx2m3-binutils-2.35.1/bin/ld: final link failed: No space left on device
make[3]: *** [../scripts/Makefile.modfinal:78: drivers/net/wireless/intel/iwlwifi/mvm/iwlmvm.ko] Error 1
/nix/store/fp2h123793bxln4cjxy0mz08sf4nx2m3-binutils-2.35.1/bin/ld: final link failed: No space left on device
make[3]: *** [../scripts/Makefile.modfinal:78: drivers/net/wireless/intel/iwlwifi/iwlwifi.ko] Error 1
make[2]: *** [../scripts/Makefile.modpost:152: __modpost] Error 2
make[1]: *** [/tmp/nix-build-linux-5.12.19.drv-0/linux-5.12.19/Makefile:1450: modules] Error 2
make: *** [../Makefile:215: __sub-make] Error 2
builder for '/nix/store/rjxmmqq2rmp8g2pm8zw05mjji47c1m3q-linux-5.12.19.drv' failed with exit code 2
note: build failure may have been caused by lack of free disk space
error: build of '/nix/store/rjxmmqq2rmp8g2pm8zw05mjji47c1m3q-linux-5.12.19.drv' failed

```

**Comment by Mic92** [comment]
@Izorkin  your error looks like you run out of disk-space. Make sure TMPDIR is not pointing to a tmpfs.

**Comment by Mic92** [comment]
@lopsided98 see https://github.com/NixOS/nixpkgs/pull/139710

**Comment by Izorkin** [comment]
I checked this variant yesterday - error build.

There is enough free space on `/tmp`.
```
df -h /tmp
Filesystem      Size  Used Avail Use% Mounted on
tmpfs            16G  4.0K   16G   1% /tmp
```

**Comment by Mic92** [comment]
The error message clearly indicates no space left errors though:

```
make[3]: *** [../scripts/Makefile.modfinal:78: drivers/net/wireless/intel/iwlegacy/iwl3945.ko] Error 1
/nix/store/fp2h123793bxln4cjxy0mz08sf4nx2m3-binutils-2.35.1/bin/ld: final link failed: No space left on device
make[3]: *** [../scripts/Makefile.modfinal:78: drivers/net/wireless/intel/ipw2x00/libipw.ko] Error 1
/nix/store/fp2h123793bxln4cjxy0mz08sf4nx2m3-binutils-2.35.1/bin/ld: final link failed: No space left on device
make[3]: *** [../scripts/Makefile.modfinal:78: drivers/net/wireless/broadcom/brcm80211/brcmfmac/brcmfmac.ko] Error 1
make[3]: *** [../scripts/Makefile.modfinal:78: drivers/net/wireless/broadcom/b43/b43.ko] Error 1
/nix/store/fp2h123793bxln4cjxy0mz08sf4nx2m3-binutils-2.35.1/bin/ld: drivers/net/wireless/ath/ath9k/ath9k_htc.ko: final close failed: No space left on device
make[3]: *** [../scripts/Makefile.modfinal:78: drivers/net/wireless/ath/ath9k/ath9k_htc.ko] Error 1
/nix/store/fp2h123793bxln4cjxy0mz08sf4nx2m3-binutils-2.35.1/bin/ld: final link failed: No space left on device
make[3]: *** [../scripts/Makefile.modfinal:78: drivers/net/wireless/intersil/hostap/hostap_cs.ko] Error 1
/nix/store/fp2h123793bxln4cjxy0mz08sf4nx2m3-binutils-2.35.1/bin/ld: final link failed: No space left on device
```

Notice that if TMPDIR can also point to `/run/user/$UID` if you are in a nix-shell, which can be surprising.

**Comment by Izorkin** [comment]
Checked it out. The build uses all 16GB of memory.
Without etnable BTF - 3,5GB.

**Comment by Mic92** [comment]
Maybe but don't build it on tmpfs than.

**Comment by lovesegfault** [comment]
I don't think we should be enabling these by default. People who want it can set `options.debug` like before.

**Comment by roblabla** [comment]
Those should definitely be enabled by default, as they enable eBPF CO-RE functionality: https://facebookmicrosites.github.io/bpf/blog/2020/02/19/bpf-portability-and-co-re.html

This is very important for a bunch of bpf-based tools. It's enabled by default on many distribs, like Fedora, Ubuntu and Debian for this reason.

**Comment by alyssais** [comment]
This ballooned the size of zfs.ko.xz ten-fold for me, from 881K to 9.4M.  Other modules from ZFS are also affected, and the combined size increase means my initramfs no longer fits on my boot partition.  But it didn't increase the size of Linux itself at all, so presumably there's some change we can make to the ZFS package to fix this? 

I do agree it makes sense for this to be enabled by default.

**Comment by Mic92** [comment]
Kernel without btf: 

```
$ wc -c bzImage
6360032
```

Kernel with btf
```
$ wc -c ./result/bzImage
7280640 ./result/bzImage
```

So it's a 14% increase for the kernel image.

**Comment by Atemu** [comment]
https://github.com/NixOS/nixpkgs/pull/141284

**Pull request merged**
